### PR TITLE
Vouchers for user promotion - Part 1: Introduction of Vouchers

### PR DESCRIPTION
**Pull request by fosterfarrell9**
In this PR, we change the way in which users can become tutors, editors or teachers. Until now, they can be selected from a dropdown. With the change implemented in this PR, **vouchers** are issued, which can then be communicated externally (e.g. in persona or mail) to the people who are supposed to get the promotion in user status, and they can then in turn redeem these vouchers in their user profile to get the desired position (see part 2 of this PR: #671). The philosophy behind this is to further reduce the amount data that can be seen by users in elevated positions.

This should fit in nicely into the existing workflow where mails are sent to tutors anyways for information exchange. Adding the voucher code there is not too much of an effort. Note that tutors could of course go ahead and share the voucher with other people, we cannot control that. But if somebody really does that, the teacher will notice and can intervene.

Because of its size, this feature is split into two PRs. This is the first one which deals with the creation and destruction of vouchers. The second one deals with the actual redemption part, see #671.

### Out of scope
This PR also
- introduces the Cypress capability to test the clipboard content via the new Clipboard API, see #684 
- automatically opens the dev tools in Cypress tests by default

### ToDo

- [x] Add Cypress Tests

### Notes for reviewers
- [x] In order to make the cypress tests run through, #683 needs to be merged.
- [x] Please check that generic users cannot create/destroy vouchers. This is covered in the unit tests (in `vouchers_controller_spec.rb`), but since this is security relevant, it would be good to check this in detail.

### Preview
image

### CODE REVIEW

File: spec/factories/vouchers.rb
Comment (Splines) [inline spec/factories/vouchers.rb:17]: We don't have a trait `:speaker` here. Lines like `build(:voucher, :speaker, lecture: seminar)` only work because `role` is the first parameter.

File: app/controllers/vouchers_controller.rb
Comment (Splines) [inline app/controllers/vouchers_controller.rb:3]: In addition to the custom `authorize! :create, @voucher`, can't we also use the `authorize_resource` here for `create`? Since we know which people are allowed to even call the endpoint. Therefore, we wouldn't have to unnecessarily create a new voucher and try to save it just to notice: "ups, you are not even allowed to call this method".

File: app/controllers/vouchers_controller.rb
Comment (Splines) [inline app/controllers/vouchers_controller.rb:2]: Instead of this, I've tried to call `set_voucher` directly at the beginning of the `invalidate` method. However, the last controller test is failing in this case as the voucher gets invalidated even by the general user. Do you know why this is the case?

**Review by Splines — CHANGES_REQUESTED** [review]
This PR looks great: well-organized and modularized, it was really fun to review!

I took the freedom to redesign the UI and also refactored the tests a bit. Just browse through the [commit page](https://github.com/MaMpf-HD/mampf/pull/670/commits) to see the changes.

File: config/routes.rb
Comment (fosterfarrell9) [reply config/routes.rb:818]: The `vouchers#invalidate` action in the vouchers controller updates the voucher in the database. It is not idempotent, which from my understanding is one difference between `GET` and `POST`, and therefore I would say it is a `POST` request. Please correct me if I am wrong.

File: app/models/voucher.rb
Comment (fosterfarrell9) [reply app/models/voucher.rb:22]: Yes, I think we do not need this in the GUI (and anywhere else). Still, I like things to be deterministic. Should we ever be in the need to look something up in a Rails console, I would like to have that something like Voucher.last acts similarly to the case of models with standards IDs and returns what is expected (and not something randomly determined).

File: db/schema.rb
Comment (fosterfarrell9) [reply db/schema.rb:13]: Can I leave that to you to make the change? This will also affect the follow-up PR, where the migrations still have the old date. They would then be performed before this migration, which would probably raise an exception since they assume that there already is a vouchers table (they refer to it). So, the dates for the migrations in #671 will need to be changed as well.

File: app/controllers/application_controller.rb
Comment (fosterfarrell9) [reply app/controllers/application_controller.rb:115]: It turned out that in the newly added controller tests for the vouchers controller, `request.session_options[:id ]` was nil, so an exception was raised. I introduced this line to get rid of the exception.

File: app/models/voucher.rb
Comment (fosterfarrell9) [reply app/models/voucher.rb:41]: I am not really sure in which order the validations corresponding to
`belongs_to :lecture`
and
`before_create :ensure_no_other_active_voucher`
are evaluated. If they are run in the order that I just used, then everything should be fine without the `return unless lecture` statement because a missing lecture would result in an error that the lecture is missing (with a proper error message). If they are run the other way around and the lecture is missing, then without the `return unless lecture` statement there would be an exception of the form "undefined method voucher for nil" raised by the next line in `ensure_no_other_voucher`. With the `return unless lecture` statement, nothing is returned and the exception is raised by the validation for `belongs_to :lecture` (with the proper error message). So, yes, in that sense this is intended. But I am not overly attached to this, if you think this is not useful then we can take that line out.

File: spec/factories/vouchers.rb
Comment (fosterfarrell9) [reply spec/factories/vouchers.rb:17]: I added the missing trait.

File: app/controllers/vouchers_controller.rb
Comment (fosterfarrell9) [reply app/controllers/vouchers_controller.rb:3]: The problem is that whether a user can create a voucher also depends on the lecture (in a lecture where the user is e.g. an editor, the user can create a voucher, but not in a lecture where the user has no additional permissions).
Therefore it states in the corresponding ability that
```
can [:create, :invalidate], Voucher do |voucher|
    user.can_update_personell?(voucher.lecture)
end
```
In order for that to work, `@voucher` in the controller needs to be set up correctly, in particular with the correct lecture. cancancan hooks right into before the create action (but after the before_actions). But `@voucher`  is only set within the create action. Therefore, with the current code, removing `except :create` and removing the manual authorization in the create action would result in anybody being able to create vouchers. If we would want to remove `except: :create` and the manual authorization for the create action, we would have to define a before action that correctly sets up `@voucher`:

```
before_action :set_new_voucher, only: :create

def set_new_voucher
  @voucher = Voucher.new(voucher_params)
end
```

File: app/controllers/vouchers_controller.rb
Comment (fosterfarrell9) [reply app/controllers/vouchers_controller.rb:2]: The reason is pretty much the same as in the above comment: If you place the call to `set_voucher` in the invalidate method and remove the before_action, then  `@voucher` is not defined in the moment when cancancan does its job,
namely after the before actions but before the invalidate action in this case.

File: config/routes.rb
Comment (Splines) [reply config/routes.rb:818]: Ah true, no you're right. `GET` shouldn't be used for operations that cause side-effects. And `POST` means we want to change something, which is the case here. Also, as you say, invalidating the voucher is of course not idempotent. So we should definitely use `POST`, don't know why I wanted this to change to `GET` to be honest 😅

File: app/models/voucher.rb
Comment (Splines) [reply app/models/voucher.rb]: Ah ok, I forgot that we use UUIDs here, so indeed it'd be nice to have some ordering in the console. I've switched over to the symbol `:created_at` in 6bc7dbb5c90.

File: db/schema.rb
Comment (Splines) [reply db/schema.rb]: Sure, done in 02a4e10f656. Will look for it in the second part of the PR as well.

File: app/controllers/application_controller.rb
Comment (Splines) [reply app/controllers/application_controller.rb:115]: Ok, I can reproduce that the tests are failing if we remove the line. However, I'm not quite sure I understand how this `store_interaction` mechanism works (the method where you added the line). Do we log every single interaction (i.e. request that the browser of the user makes) via a Sidekiq worker in the interactions database? If so, why was this introduced (to prevent abuse of MaMpf?) and what's its relationship to GDPR?

File: app/models/voucher.rb
Comment (Splines) [reply app/models/voucher.rb:41]: I'm not too sure either based on some digging in the documentation, so I posted [this question on StackOverflow](https://stackoverflow.com/q/78959006/9655481). Let's see if we get an answer there.

For now, we can leave the `return unless lecture` in here as safeguard.

File: app/controllers/vouchers_controller.rb
Comment (Splines) [reply app/controllers/vouchers_controller.rb]: As discussed off-GitHub, there's a [`load_and_authorize_resource`](https://github.com/CanCanCommunity/cancancan/blob/develop/docs/controller_helpers.md#authorize_resource-load_resource-load_and_authorize_resource) method that simplifies this. Added in cb5863dd.

File: app/controllers/vouchers_controller.rb
Comment (Splines) [reply app/controllers/vouchers_controller.rb]: See my other comment and commit cb5863dda.

File: app/controllers/application_controller.rb
Comment (Splines) [reply app/controllers/application_controller.rb:115]: discussed off-GitHub, it is compliant

File: app/models/voucher.rb
Comment (Splines) [reply app/models/voucher.rb:41]: @fosterfarrell9 We got a nice answer [here](https://stackoverflow.com/a/78959761/9655481) that argues with the actual Rails soruce code. Based on this, I will just leave the `return unless lecture` in for safety, even though it's almost certain that we would not run into any problems if omitted.

**Review by Splines — DISMISSED** [review]
LGTM now! I've added some more cypress tests to also test the expiration of vouchers there (via newly introduced time travelling).

Please make sure to review everything I added to this PR by means of my last commits before finally **squashing and merging** into the `dev` branch.

After the "squash and merge", I will take care of merge conflicts in the second part of the pull request.

**Review by Splines — APPROVED** [review]
the only thing left to do now is resolve the one conversation and merge dev into this branch where I fixed the flaky unit test, then rerun the tests and squash and merge

**Pull request merged**
